prow/gerrit/reporter/reporter.go

```diff
@@ -103,46 +111,65 @@ func (c *Client) ShouldReport(pj *v1.ProwJob) bool {
 	// If you are hitting here, which means the entire patchset has been finished :-)
 
+	logger := logrus.WithField("prowjob", pj)
+
 	clientGerritRevision := client.GerritRevision
 	clientGerritID := client.GerritID
 	clientGerritInstance := client.GerritInstance
 	pjTypeLabel := kube.ProwJobTypeLabel
+	gerritReportLabel := client.GerritReportLabel
 
-	// list all prowjobs in the patchset matching pj's type (pre- or post-submit)
 	selector := labels.Set{
 		clientGerritRevision: pj.ObjectMeta.Labels[clientGerritRevision],
 		pjTypeLabel:          pj.ObjectMeta.Labels[pjTypeLabel],
 	}
-	pjsOnRevision, err := c.lister.List(selector.AsSelector())
+	if pj.ObjectMeta.Labels[gerritReportLabel] == "" {
+		// Shouldn't happen, adapter should already have defaulted to Code-Review
+		logger.Errorf("Gerrit report label not set for job %s", pj.Spec.Job)
+	} else {
+		selector[gerritReportLabel] = pj.ObjectMeta.Labels[gerritReportLabel]
+	}
+
+	// list all prowjobs in the patchset matching pj's type (pre- or post-submit)
+
+	pjsOnRevisionWithSameLabel, err := c.lister.List(selector.AsSelector())
 	if err != nil {
-		logrus.WithError(err).Errorf("Cannot list prowjob with selector %v", selector)
-		return err
+		logger.WithError(err).Errorf("Cannot list prowjob with selector %v", selector)
+		return nil, err
 	}
 
 	// generate an aggregated report:
 	total := 0
 	success := 0
 	message := ""
-
-	for _, pjOnRevision := range pjsOnRevision {
-		if pjOnRevision.Status.PrevReportStates[c.GetName()] == pjOnRevision.Status.State {
-			logrus.Infof("Revision %s has been reported already", pj.ObjectMeta.Labels[clientGerritRevision])
-			return nil
+	var toReportJobs []*v1.ProwJob
+	mostRecentJob := map[string]*v1.ProwJob{}
+	for _, pjOnRevisionWithSameLabel := range pjsOnRevisionWithSameLabel {
+		job, ok := mostRecentJob[pjOnRevisionWithSameLabel.Spec.Job]
+		if !ok || job.CreationTimestamp.Time.Before(pjOnRevisionWithSameLabel.CreationTimestamp.Time) {
+			mostRecentJob[pjOnRevisionWithSameLabel.Spec.Job] = pjOnRevisionWithSameLabel
 		}
-
-		if pjOnRevision.Status.State == v1.AbortedState {
+	}
+	for _, pjOnRevisionWithSameLabel := range mostRecentJob {
+		if pjOnRevisionWithSameLabel.Status.State == v1.AbortedState {
 			continue
 		}
-
+		toReportJobs = append(toReportJobs, pjOnRevisionWithSameLabel)
 		total++
-		if pjOnRevision.Status.State == v1.SuccessState {
+		if pjOnRevisionWithSameLabel.Status.State == v1.SuccessState {
 			success++
 		}
 
-		message = fmt.Sprintf("%s\nJob %s finished with %s -- URL: %s", message, pjOnRevision.Spec.Job, pjOnRevision.Status.State, pjOnRevision.Status.URL)
+		message = fmt.Sprintf("%s\nJob %s finished with %s -- URL: %s", message, pjOnRevisionWithSameLabel.Spec.Job, pjOnRevisionWithSameLabel.Status.State, pjOnRevisionWithSameLabel.Status.URL)
+	}
+
+	if total <= 0 {
+		// Shouldn't happen but return if does
+		logger.Warn("Tried to report empty or aborted jobs.")
```

Comment:
Requested from users :man_shrugging:

Reply:
What was requested from users?
To clarify, I'm asking why the comment is valid. AFAICT it is possible for this to happen if there is only one job in the patchset and it is aborted.

Reply:
Request from users is to not report aborted jobs since it generates spam. To accomodate that request we now skip reporting if the job is aborted https://github.com/kubernetes/test-infra/pull/11448/files/1fd22c2917d0a89bb496bc7483a4469dbfe3676c#diff-44b62b4ebd1dd28fcdde3dd0649b0140R157. This entire `total <= 0` block is for the exact case that you mentioned :) otherwise it would go ahead and report a `0 of 0 jobs passed` message.

Reply:
@cjwagner so remember gerrit doesn't allow edit comments and doesn't have context, what happened before:

- user uploaded patchset 1
- prow comment 1: job foo is triggered for patchset 1
- user uploaded patchset 2
- prow comment 2: job foo is triggered for patchset 2
- prow comment 3: job foo is finished at aborted state for patchset 1, and vote -1 on code-review (this happens after plank abort the first prowjob)

And note that the comment 3 was confusing and not useful because user doesn't know why the job was aborted.